File: leetcode/q679.java
package leetcode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class q679 {
    class Solution {
        public boolean judgePoint24(int[] cards) {
            List<Integer> arr = new ArrayList<>();
            for(int card: cards) arr.add(card);
            return helper(arr);
        }

        private boolean helper(List<Integer> arr) {
            if(arr.size() == 1) {
                if(arr.get(0) == 24) return true;
                return false;
            }
            
            boolean res = false;
            for(int i = 0; i < arr.size(); i++) {
                for(int j = 0; j < i; j++) {
                    List<Integer> next = new ArrayList<>();
                    int card1 = arr.get(i);
                    int card2 = arr.get(j);

                    next.addAll(Arrays.asList(card1+card2, card1-card2, card2-card1, card1*card2));
                    if(card1 > 0 && card2%card1 == 0) next.add(card2/card1);
                    if(card2 > 0 && card1%card2 == 0) next.add(card1/card2);

                    arr.remove(i);
                    arr.remove(j);

                    for(Integer card: next) {
                        arr.add(card);
                        res = res || helper(arr);
                        arr.remove(arr.size()-1);
                    }

                    arr.add(j, card2);
                    arr.add(i, card1);
                }
            }

            return res;
        }
    }
}

File: leetcode/q679.js
/**
 * @param {number[]} cards
 * @return {boolean}
 */
var judgePoint24 = function(cards) {
    if(cards.length == 1) {
        if(cards[0] == 24) return true;
        return false;
    }

    var res = false;
    for(let i = 0; i < cards.length; i++) {
        for(let j = 0; j < i; j++) {
            let newCards = [];
            newCards.push(cards[i]+cards[j], cards[i]-cards[j], cards[j]-cards[i], cards[i]*cards[j]);
            if(cards[i] != 0 && cards[j]%cards[i] == 0) newCards.push(cards[j]/cards[i]);
            if(cards[j] != 0 && cards[i]%cards[j] == 0) newCards.push(cards[i]/cards[j]);

            let newList = [];
            for(let k = 0; k < cards.length; k++) {
                if(k != i && k != j) {
                    newList.push(cards[k]);
                }
            }
            
            for(card of newCards) {
                newList.push(card);
                res = res || judgePoint24(newList);
                newList.pop();
            }
        }
    }

    return res;
};